mono/metadata/w32socket.c

```diff
@@ -2812,15 +2631,52 @@ mono_network_cleanup (void)
 	mono_networking_shutdown ();
 }
 
+/*
+ * This is a very specific function whose only purpose is to
+ * break a given thread from socket syscalls.
+ *
+ * This only exists because linux won't fail a call to connect
+ * if the underlying is closed.
+ *
+ * TODO We should cleanup and unify this with the other syscall abort
+ * facility.
+ */
 void
 ves_icall_cancel_blocking_socket_operation (MonoThreadObjectHandle thread, MonoError *error)
```

Comment:
I imagine mono_thread_info_abort_socket_syscall_for_close was in threads.c because it is duplicating so much of its gnarly internals. Perhaps that should not be changed?

But then, more so, can this instead be more like:

```
static SuspendThreadResult
mono_thread_abort_syscall_callback (THREAD_INFO_TYPE *info, gpointer user_data)
{
	return MonoResumeThread;
}

void	
mono_thread_abort_syscall (MonoNativeThreadId tid)	
{
	if (tid == mono_native_thread_id_get ())
		return;
	mono_thread_info_safe_suspend_and_run (id, TRUE/*interrupt kernel*/, mono_thread_abort_syscall_callback, NULL);
}
```
?

Or, why is this not just calling ves_icall_System_Threading_Thread_Interrupt_internal ()?
With an intervening mono_thread_set_state/mono_thread_clr_state (thread, ThreadState_WaitSleepJoin);
Why is it so different than other interruptable kernel waits or syscalls?

The comment about Linux seems off.
It is only common sense that a close on one thread will not interrupt
the kernel on another when it has already taken a reference to the underlying data.
Doing a close to implement this would be dangerous.

Nobody tries to interrupt other waits or I/O with  close, so should they not interrupt socket functions that way either.